src/uu/pgrep/src/pgrep.rs

```diff
@@ -0,0 +1,216 @@
+// This file is part of the uutils procps package.
+//
+// For the full copyright and license information, please view the LICENSE
+// file that was distributed with this source code.
+
+pub mod pid;
+
+use clap::{arg, crate_version, Arg, ArgAction, ArgGroup, ArgMatches, Command};
+use pid::{walk_pid, PidEntry};
+use std::{borrow::BorrowMut, cmp::Ordering};
+use uucore::{error::UResult, format_usage, help_about, help_usage};
+
+const ABOUT: &str = help_about!("pgrep.md");
+const USAGE: &str = help_usage!("pgrep.md");
+
+#[uucore::main]
+pub fn uumain(args: impl uucore::Args) -> UResult<()> {
+    let matches = uu_app().try_get_matches_from(args)?;
+    let patterns = collect_arg_patterns(&matches);
+
+    // Some(()) => Detected
+    // None => Go on
+    if handle_oldest_newest(&matches, &patterns).is_some() {
+        return Ok(());
+    }
+
+    handle_normal_pid_collect(&matches, &patterns);
+
+    Ok(())
+}
+fn collect_arg_patterns(matches: &ArgMatches) -> Vec<String> {
+    let should_ignore_case = matches.get_flag("ignore-case");
+
+    let programs = matches
+        .get_many::<String>("pattern")
+        .unwrap_or_default()
+        .map(|it| it.to_string());
+
+    if should_ignore_case {
+        programs.map(|it| it.to_lowercase()).collect::<Vec<_>>()
+    } else {
+        programs.collect()
+    }
+}
+
+fn collect_pid(matches: &ArgMatches, patterns: &[String]) -> Vec<PidEntry> {
+    let should_inverse = matches.get_flag("inverse");
+    let should_ignore_case = matches.get_flag("ignore-case");
+
+    walk_pid()
+        .filter(move |it| {
+            let binding = it.to_owned().borrow_mut().status();
+            let Some(name) = binding.get("Name") else {
+                return false;
+            };
+
+            // Processs flag `--ignore-case`
+            let name = if should_ignore_case {
+                name.to_lowercase()
+            } else {
+                name.into()
+            };
+
+            patterns.iter().any(|it| name.contains(it)) ^ should_inverse
+        })
+        .collect::<Vec<_>>()
+}
+
+// Make -o and -n as a group of args
+fn handle_oldest_newest(matches: &ArgMatches, patterns: &[String]) -> Option<()> {
```

Comment:
The `Option<()>` return type is a hack, it's a disguised bool ;-)  I think it would make more sense to return a `Option<String>`, with the string being a pid. And doing the printing outside of this function.

Reply:
Maybe it's better to evaluate result string inside the function to return an `Option<String>`, and print it outside the function?

Reply:
Yes, probably.